source/common/http/http1/codec_impl.cc

```diff
@@ -611,9 +617,31 @@ int ConnectionImpl::onHeadersCompleteBase() {
   return handling_upgrade_ ? 2 : rc;
 }
 
+void ConnectionImpl::bufferBody(const char* data, size_t length) {
+  buffered_body_.add(data, length);
+}
+
+void ConnectionImpl::dispatchBufferedBody() {
+  ASSERT(HTTP_PARSER_ERRNO(&parser_) == HPE_OK || HTTP_PARSER_ERRNO(&parser_) == HPE_PAUSED);
+  if (buffered_body_.length() > 0) {
+    onBody(buffered_body_);
+    buffered_body_.drain(buffered_body_.length());
+  }
+}
+
+void ConnectionImpl::onChunkHeader(bool is_final_chunk) {
+  if (is_final_chunk) {
+    // Dispatch body before parsing trailers, so body ends up dispatched even if an error is found
+    // while processing trailers.
+    dispatchBufferedBody();
```

Comment:
lazy ask: we have a test for this?

Reply:
Http1ServerConnectionImplTest.Http11InvalidTrailerPost

I'm not 100% sure if this is actually necessary.  If we find a trailer error, an exception is throw even if trailer proxying is disabled.  Pushing the body or throwing it away when trailer processing fails seems about equally good.  Thoughts?

Reply:
Mainly I wanted to make sure we regression tested that all body was sent before trailers in the happy path.  I don't have strong opinions about making sure all data gets processed in case of errors.

Reply:
I enhanced Http1ServerConnectionImplTest.RequestWithTrailersKept to verify that body is delivered before trailers.

The call to dispatchBufferedBody() in ConnectionImpl::onMessageCompleteBase is enough to accomplish that.